src/log.c

```diff
@@ -16,527 +18,391 @@
-    freeRawLogEntry(re);
 
     /* Read Entries */
-    do {
-        raft_entry_t *e = NULL;
-
-        long offset = ftell(log->file);
-        if (readRawLogEntry(log->file, &re) < 0 || !re->num_elements) {
-            break;
-        }
-
-        if (!strcasecmp(re->elements[0].ptr, "ENTRY")) {
-            e = parseRaftLogEntry(re);
-            if (!e) {
-                freeRawLogEntry(re);
-                ret = -1;
-                break;
+    while (true) {
+        uint64_t offset = (uint64_t) FileGetReadOffset(&log->file);
+        raft_entry_t *e = readEntry(log);
+        if (!e) {
+            size_t bytes = FileSize(&log->file) - offset;
+            if (bytes > 0) {
+                LOG_WARNING("Found partial entry at the end of the log file. "
+                            "Discarding %zu bytes.",
+                            bytes);
+
+                int rc = FileTruncate(&log->file, offset);
+                RedisModule_Assert(rc == RR_OK);
             }
-            log->index++;
-            ret++;
 
-            updateIndex(log, log->index, offset);
-        } else {
-            LOG_WARNING("Invalid log entry: %s", (char *) re->elements[0].ptr);
-            freeRawLogEntry(re);
-
-            ret = -1;
-            break;
+            return RR_OK;
         }
-
-        freeRawLogEntry(re);
         raft_entry_release(e);
-    } while (1);
 
-    if (ret > 0) {
-        log->num_entries = ret;
+        log->index++;
+        log->num_entries++;
+
+        ssize_t rc = FileWrite(&log->idxfile, &offset, sizeof(offset));
+        RedisModule_Assert(rc == sizeof(offset));
     }
-    return ret;
 }
 
-RRStatus RaftLogWriteEntry(RaftLog *log, raft_entry_t *entry)
+size_t LogFileSize(Log *log)
 {
-    size_t written = 0;
-    int n;
-
-    if ((n = writeBegin(log->file, 5)) < 0) {
-        return RR_ERROR;
-    }
-    written += n;
-    if ((n = writeBuffer(log->file, "ENTRY", 5)) < 0) {
-        return RR_ERROR;
-    }
-    written += n;
-    if ((n = writeUnsignedInteger(log->file, entry->term, 0)) < 0) {
-        return RR_ERROR;
-    }
-    written += n;
-    if ((n = writeUnsignedInteger(log->file, entry->id, 0)) < 0) {
-        return RR_ERROR;
-    }
-    written += n;
-    if ((n = writeUnsignedInteger(log->file, entry->type, 0)) < 0) {
-        return RR_ERROR;
-    }
-    written += n;
-    if ((n = writeBuffer(log->file, entry->data, entry->data_len)) < 0) {
-        return RR_ERROR;
-    }
-    written += n;
-
-    /* Update index */
-    log->file_size = ftell(log->file);
-    off_t offset = log->file_size - written;
-    log->index++;
-    updateIndex(log, log->index, offset);
-
-    return RR_OK;
+    return FileSize(&log->file);
 }
 
-RRStatus RaftLogSync(RaftLog *log, bool sync)
+int LogSync(Log *log, bool sync)
 {
     uint64_t begin = RedisModule_MonotonicMicroseconds();
 
-    if (fflush(log->file) < 0) {
-        LOG_WARNING("fflush() : %s", strerror(errno));
+    if (FileFlush(&log->file) != RR_OK) {
```

Comment:
Added log lines. 

Actually, caller should handle these errors as these are super important errors. e.g. Disk is full. Currently, it is missing. 
This can be the last step of our effort: Handling disk full error, applying backpressure (rejecting new entries until we replicate & apply existing ones, otherwise we can fill the disk if the new ones are coming faster then we consume).